certstore_linux.go

```diff
@@ -1,14 +1,321 @@
 
-// This will hopefully give a compiler error that will hint at the fact that
-// this package isn't designed to work on Linux.
-func init() {
-	CERTSTORE_DOESNT_WORK_ON_LINIX
+#include <nss.h>
+#include <cryptohi.h>
+#include <nspr.h>
+#include <p12.h>
+#include <p12plcy.h>
+#include <pk11pub.h>
+#include <prerror.h>
+#include <secerr.h>
+
+SECItem *P12U_NicknameCollisionCallback(SECItem *old_nick, PRBool *cancel, void *wincx) {
+	char *nick = NULL;
+	SECItem *ret_nick = NULL;
+	CERTCertificate *cert = (CERTCertificate *)wincx;
+	if (!cancel || !cert) {
+		return NULL;
+	}
+	nick = CERT_MakeCANickname(cert);
+	if (!nick) {
+		return NULL;
+	}
+	if (old_nick && old_nick->data && old_nick->len &&
+		PORT_Strlen(nick) == old_nick->len - 1 &&
+		!PORT_Strncmp((char *)old_nick->data, nick, old_nick->len - 1)) {
+		PORT_Free(nick);
+		return NULL;
+    }
+	ret_nick = PORT_ZNew(SECItem);
+	if (ret_nick == NULL) {
+		PORT_Free(nick);
+		return NULL;
+	}
+	ret_nick->data = (unsigned char *)nick;
+	ret_nick->len = PORT_Strlen(nick) + 1;
+	return ret_nick;
+}
+
+CERTCertListNode *CertListHead(CERTCertList *l) {
+	return CERT_LIST_HEAD(l);
+}
+
+CERTCertListNode *CertListNext(CERTCertListNode *n) {
+	return CERT_LIST_NEXT(n);
+}
+
+int CertListEnd(CERTCertListNode *n, CERTCertList *l) {
+	return CERT_LIST_END(n, l);
+}
+
+const char *GetErrorString() {
+	return PORT_ErrorToString(PORT_GetError());
+}
+
+void EnablePKCS12Algorithms() {
+	SEC_PKCS12EnableCipher(PKCS12_RC4_40, 1);
+	SEC_PKCS12EnableCipher(PKCS12_RC4_128, 1);
+	SEC_PKCS12EnableCipher(PKCS12_RC2_CBC_40, 1);
+	SEC_PKCS12EnableCipher(PKCS12_RC2_CBC_128, 1);
+	SEC_PKCS12EnableCipher(PKCS12_DES_56, 1);
+	SEC_PKCS12EnableCipher(PKCS12_DES_EDE3_168, 1);
+#ifdef PKCS12_AES_CBC_128
+	SEC_PKCS12EnableCipher(PKCS12_AES_CBC_128, 1);
+	SEC_PKCS12EnableCipher(PKCS12_AES_CBC_192, 1);
+	SEC_PKCS12EnableCipher(PKCS12_AES_CBC_256, 1);
+#endif
+	SEC_PKCS12SetPreferredCipher(PKCS12_DES_EDE3_168, 1);
+}
+*/
+import "C"
+import (
+	"crypto"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"unicode/utf16"
+	"unsafe"
+)
+
+// Map go hash algorithm identifiers to NSS's SECOidTags
+var goToNssAlg = map[crypto.Hash]C.SECOidTag{
+	crypto.SHA1:   C.SEC_OID_SHA1,
+	crypto.SHA256: C.SEC_OID_SHA256,
+	crypto.SHA384: C.SEC_OID_SHA384,
+	crypto.SHA512: C.SEC_OID_SHA512,
+}
+
+// nssStore is a wrapper around a C.NSSInitContext
+type nssStore C.NSSInitContext
+
+// Identities implements the Store interface.
+func (store nssStore) Identities() ([]Identity, error) {
+	var (
+		identities = make([]Identity, 0)
+		certs      = C.PK11_ListCerts(C.PK11CertListUser, nil)
+		node       *C.CERTCertListNode
+	)
+	if certs == nil {
+		return nil, fmt.Errorf("error listing user certificates: %s", C.GoString(C.GetErrorString()))
+	}
+	defer C.CERT_DestroyCertList(certs)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*C.CERT_DupCertificate(node.cert))
```

Comment:
Similar problem here, I think. Could this be `(*nssIdentity)(C.CERT_DupCertificate(node.cert))` to avoid copying the struct?

Reply:
Maybe. I think I tried something like that and got shouted at by cgo but I'll have a play.